core/src/main/scala/org/http4s/client/jdkhttpclient/JdkWebSocketClient.scala

```diff
@@ -0,0 +1,129 @@
+                toCompletionStage(for {
+                  _ <- queue.enqueue1(wsf.some)
+                  // if we encounter an error or receive a Close frame, we close the queue
+                  _ <- wsf match {
+                    case Left(_) | Right(_: WSFrame.Close) => queue.enqueue1(none)
+                    case _ => F.unit
+                  }
+                } yield ())
+              wsListener = new JWebSocket.Listener {
+                override def onOpen(webSocket: JWebSocket): Unit = ()
+                override def onClose(webSocket: JWebSocket, statusCode: Int, reason: String)
+                    : CompletionStage[_] =
+                  // The output side of this connection will be closed when the returned CompletionStage completes.
+                  // Therefore, we return a never completing CompletionStage, so we can control when the output will
+                  // be closed (as it is allowed to continue sending frames (as few as possible) after a close frame
+                  // has been received).
+                  handleReceive(WSFrame.Close(statusCode, reason).asRight)
+                    .thenCompose[Nothing](new Function[Unit, CompletionStage[Nothing]] {
+                      override def apply(t: Unit) = new CompletableFuture[Nothing]
+                    })
+                override def onText(webSocket: JWebSocket, data: CharSequence, last: Boolean)
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Text(data.toString, last).asRight)
+                override def onBinary(webSocket: JWebSocket, data: ByteBuffer, last: Boolean)
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Binary(ByteVector(data), last).asRight)
+                override def onPing(webSocket: JWebSocket, message: ByteBuffer)
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Ping(ByteVector(message)).asRight)
+                override def onPong(webSocket: JWebSocket, message: ByteBuffer)
+                    : CompletionStage[_] =
+                  handleReceive(WSFrame.Pong(ByteVector(message)).asRight)
+                override def onError(webSocket: JWebSocket, error: Throwable): Unit = {
+                  handleReceive(error.asLeft); ()
+                }
+              }
+              webSocket <- fromCompletableFuture(
+                F.delay(wsBuilder.buildAsync(URI.create(uri.renderString), wsListener))
+              )
+              sendSem <- Semaphore[F](1L)
+            } yield (webSocket, queue, sendSem)
+          } {
+            case (webSocket, _, _) =>
+              for {
+                isOutputOpen <- F.delay(!webSocket.isOutputClosed)
+                closeOutput = fromCompletableFuture(
+                  F.delay(webSocket.sendClose(JWebSocket.NORMAL_CLOSURE, ""))
+                )
+                _ <- closeOutput.whenA(isOutputOpen)
+                _ <- F.delay(webSocket.abort())
```

Comment:
I am not sure if closing both sides apruptly is a good idea immediately after sending a close frame, but it won't matter a lot and it protects us from a server which continues to send frames afterwards. Maybe sleep for a second beforehand?

Reply:
I don't know the web socket protocol very well, but [section 7.1.1 of RFC6455 says](https://tools.ietf.org/html/rfc6455#page-41):

> The underlying TCP connection, in most normal cases, SHOULD be closed
   first by the server, so that it holds the TIME_WAIT state and not the
   client (as this would prevent it from re-opening the connection for 2
   maximum segment lifetimes (2MSL), while there is no corresponding
   server impact as a TIME_WAIT connection is immediately reopened upon
   a new SYN with a higher seq number).  In abnormal cases (such as not
   having received a TCP Close from the server after a reasonable amount
   of time) a client MAY initiate the TCP Close.  As such, when a server
   is instructed to _Close the WebSocket Connection_ it SHOULD initiate
   a TCP Close immediately, and when a client is instructed to do the
   same, it SHOULD wait for a TCP Close from the server.

Reply:
Thx for the pointer - even though aborting immediately is allowed, it is discouraged, and closing it after a timeout is optional - I will remove the line.